### PR TITLE
Implement SipHash-2-4

### LINKED ISSUE #373: Implement SipHash-2-4
## Context

#198 proposes a typed, covered btclib equivalent of the deterministic primitives exposed by Bitcoin Core's Python functional-test framework. SipHash-2-4 is one of the clearest missing primitives:

- Core implements `siphash(k0, k1, data)` and the 256-bit convenience wrapper `siphash256(k0, k1, num)` in [`test_framework/crypto/siphash.py`](https://github.com/bitcoin/bitcoin/blob/17c5e33e9c5418fb0240f5d4210c87f54b88cdda/test/functional/test_framework/crypto/siphash.py).
- btclib currently has no SipHash implementation.
- #218 already records that Core's `src/test/data/siphash.json` is not vendored for that reason.
- SipHash is a prerequisite for BIP158 compact block filters and BIP152 compact-block short IDs.

## Proposed scope

- Implement SipHash-2-4 as a typed, deterministic btclib primitive.
- Accept an explicit 128-bit key (or two documented 64-bit words) and bytes-like input.
- Return the unsigned 64-bit result with byte order and key interpretation documented.
- Provide a 32-byte/`uint256` convenience function only if it avoids duplicated Bitcoin-specific byte-order code in BIP152/BIP158.
- Validate key width and input types using btclib's normal exception contracts.
- Keep the primitive independent of networking and node/RPC code.

The exact public module and function names should follow the existing btclib hash/crypto layout; compatibility aliases for Core's helper names are not required unless #198 later adds a dedicated adapter layer.

## Tests

- Vendor and pin Bitcoin Core's `src/test/data/siphash.json` following `tests/_data/README.md`.
- Reproduce every Core vector byte-for-byte.
- Cover empty input, partial final words, multiple eight-byte words, key boundaries and invalid inputs.
- Cross-check the 32-byte convenience form against the generic byte-oriented function.
- Preserve the project's strict typing and coverage requirements.

## Acceptance criteria

- All Core SipHash vectors pass.
- The public API documents key width, integer/byte ordering and return range.
- BIP152 and BIP158 can consume the primitive without carrying their own SipHash implementation.
- No socket, RPC or transport dependency is introduced.

Related: #198, #218.

---

Surveyed against Bitcoin Core `17c5e33e9c5418fb0240f5d4210c87f54b88cdda` and btclib `77eb4483444ae54c8622bb5efad664e83c4450f7`.

### PR DESCRIPTION
**Pull request by fametrano**
Closes #373.

## What changed

`hashes.siphash(k0, k1, octets)` is Bitcoin Core's SipHash-2-4
(`crypto/siphash.cpp`, mirrored in Python at
`test/functional/test_framework/crypto/siphash.py`), pure Python and
typed to btclib's own contracts: `k0`/`k1` are the two 64-bit words of
the 128-bit key, validated as unsigned 64-bit integers
(`BTClibTypeError`/`BTClibValueError`); `octets` takes every spelling
`Octets` already accepts; the answer is an unsigned 64-bit `int`.

It lives in `btclib/hashes.py` beside the library's other hash
functions, which `btclib/block/` already imports from directly
(`block_header.py`, `block.py`, `merkle_proof.py`) — so `btclib/block/`
(issue #374, BIP158) will import `siphash` the same way, with no new
layering to cross.

**No separate 32-byte/`uint256` convenience is added.** Bitcoin Core's
Python `siphash256(k0, k1, num)` exists only to turn a `uint256` read
as a Python `int` back into the 32 bytes `siphash` was going to hash
in the first place (`num.to_bytes(32, 'little')`). btclib already
represents a 32-byte hash as bytes in that same internal order
(`hashes.merkle_root` and its callers), so a caller here passes such a
value to `siphash` directly and there is nothing left for a wrapper to
save. This also matches what BIP158 itself needs: Core's Python
reference filter (`test_framework/blockfilter.py`) calls
`siphash(k0, k1, script_pub_key)` on the generic byte form, not the
256-bit one — `siphash256` is BIP152's helper, for hashing a `tx_hash`,
and BIP152 is not in scope here.

## Tests

Core's `src/test/data/siphash.json` — 146 vectors, empty input through
64-byte inputs, partial and multiple 8-byte words, key boundaries —
vendored whole as `tests/_data/siphash.json` (pin and blob check in
`tests/_data/README.md`) and reproduced byte for byte in
`tests/siphash_test.py`, alongside key-boundary and invalid-input
(wrong key width, wrong input type) coverage. 64 of the 146 rows also
carry `expected.siphash13uj`, Core's unrelated unpadded jumbo-block
SipHash-1-3 variant (`SipHasher13UJ`); btclib implements no such
hasher, so that field is unread.

A `CHANGELOG.md` entry is under "The public API and the module
layout".

## Verification transcript

```
$ uv run pre-commit run --all-files
...
mypy (strict, in the project environment).......................................Passed
uv-lock.........................................................................Passed
check-manifest..................................................................Passed
Check package with Pyroma.......................................................Passed

$ uv run pytest
...
TOTAL   41836      0  100.00%
Required test coverage of 100.0% reached. Total coverage: 100.00%
28416 passed, 7 skipped in 27.14s

$ uv run --locked --no-default-groups --group docs \
    sphinx-build -W --keep-going -b html docs/source docs/build/html
...
build succeeded.
The HTML pages are in docs/build/html.
$ echo $?
0
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Add Bitcoin Core-compatible SipHash-2-4 hashing to the public hashes API with comprehensive conformance tests.

New Features:
- Add a public `hashes.siphash` implementation of Bitcoin Core-compatible SipHash-2-4 for octet inputs and validated 128-bit keys.

Enhancements:
- Expose SipHash-2-4 alongside the library’s existing hash functions with btclib-specific input and error contracts.

Documentation:
- Document the new SipHash API and its intended use in the changelog.

Tests:
- Add Bitcoin Core’s complete 146-vector SipHash-2-4 test corpus, plus boundary and invalid-input coverage.

Chores:
- Vendor and pin Bitcoin Core’s SipHash test vectors in the project test data.

### CODE REVIEW

File: tests/_data/README.md [hunk 1]
```diff
@@ -1041,6 +1042,30 @@ what reads them is `base58._b58encode`/`_b58decode` and not the checked
...
+pulled  2026-08-20
+behind  0 revisions; that commit is the tip of the path
+```
+
+Verdict: **identical**, 146 cases. Each carries a key, an input split
+into blocks (so that a multi-block input and its byte-for-byte
+concatenation are the same case, on purpose), and the expected output
+of two constructions this library implements one of:
+`expected.siphash24` is standard SipHash-2-4, what `hashes.siphash`
+answers and what `tests/hashes_test.py` checks every row against, input
```
[the remark this answers is not in the document]
Comment (fametrano) [reply tests/_data/README.md]: Fixed in c15d644: the pointer now names `tests/siphash_test.py`, which is where the 146 rows are read.
